packages/labs/ssr/src/lib/render.ts

```diff
@@ -43,5 +70,96 @@ export function* render(
   if (isTemplateResult(value)) {
     hydratable = isHydratable(value);
   }
-  yield* renderValue(value, renderInfo as RenderInfo, hydratable);
+  return renderValue(value, renderInfo as RenderInfo, hydratable);
+}
+
+type InternalRenderResultIterator = Iterator<string | Thunk>;
+
+/**
+ * Wraps a ThunkedRenderResult to implement a RenderResult.
+ */
+export class RenderResultIterator
+  implements Iterator<string | Promise<RenderResult>>
+{
+  /**
+   * A stack of open iterators.
+   *
+   * As the ThunkedRenderResult thunks are called, they may yield new arrays of
+   * strings and/or thunks. We push the iterators for these arrays onto the
+   * stack and consume them depth-first.
+   */
+  private _iterators: Array<InternalRenderResultIterator>;
+  private _waiting = false;
+
+  constructor(result: ThunkedRenderResult) {
+    this._iterators = [result[Symbol.iterator]()];
+  }
+
+  next(): IteratorResult<string | Promise<RenderResult>, unknown> {
+    if (this._waiting) {
+      throw new Error(
+        'Cannot call next() while waiting for a Promise to resolve'
+      );
+    }
+    const iterator = this._iterators.at(-1);
```

Comment:
Since most JS VMs don't have tail calls, can probably get a bit of speed out of putting most of this inside a `while (true)` so we can `continue;` instead of recursing.

Reply:
True. It should be very minimal as calling `next()` will only recurse when entering or leaving a nested array or have a string of undefineds, and the stack unwinds when it returns. So you only get deep recursion if you have either a nested set of arrays that are all the first or last value of their parent. Or a string of undefined, which should be rare. Either way, I'll make the change :)